README.md
pail
====

Python Adaptive Images Library


CHANGES.txt
Changes
=======

0.3.dev0 (unreleased)
---------------------

- Drop support for Python 3.2.

- Officially support Python 3.4.

- Support retina displays.

0.2 (2013-05-17)
----------------

- Add very plain mobile detection. Now, if no device width is given
  (via cookie) detected mobile devices get images with the smallest
  predefined resolution while non-mobile devices get the images with
  the largest predefined resolution.

- Removed zc.buildout stuff (we do not need it).

0.1 (2013-05-07)
----------------

- First implementation based on Matt Wilcox' PHP script.


pail/tests/test_wsgi.py
import os
import unittest
from pail.wsgi import ImageAdaptingMiddleware, filter_app
try:
    from cStringIO import StringIO
except ImportError:                     # pragma: no cover
    from io import BytesIO as StringIO  # pragma: no cover
from PIL import Image
from webob import Request

HTML = '''\
<html>
    <body>
        <h1>Content title</h1>
        <div id="content">Content content</div>
    </body>
</html>
'''
DATA_JPEG = open(
    os.path.join(os.path.dirname(__file__), 'lena.jpg'), 'rb').read()

DATA_GIF = open(
    os.path.join(os.path.dirname(__file__), 'lena.gif'), 'rb').read()

DATA_PNG = open(
    os.path.join(os.path.dirname(__file__), 'lena.png'), 'rb').read()

DEFAULT_COOKIE = 'resolution=128; $PATH=/'

TEST_RESOLUTIONS = '128, 64, 32'


def wsgi_app_plaintext(environ, start_response):
    start_response('200 OK', [('Content-Type', 'text/plain; charset=UTF-8')])
    return ['Hi!']


def wsgi_app_html(environ, start_response):
    start_response('200 OK', [('Content-Type', 'text/html; charset=UTF-8')])
    return [HTML]


def wsgi_app_no_content_type(environ, start_response):
    start_response('200 OK', [('Content-Type', '')])
    return []


def wsgi_app_img_jpg(environ, start_response):
    start_response('200 OK', [('Content-Type', 'image/jpeg'),
                              ('Content-Length', '%s' % len(DATA_JPEG))])
    return [DATA_JPEG]


def wsgi_app_img_gif(environ, start_response):
    start_response('200 OK', [('Content-Type', 'image/gif'),
                              ('Content-Length', '%s' % len(DATA_GIF))])
    return [DATA_GIF]


def wsgi_app_img_png(environ, start_response):
    start_response('200 OK', [('Content-Type', 'image/gif'),
                              ('Content-Length', '%s' % len(DATA_PNG))])
    return [DATA_PNG]


class WSGITests(unittest.TestCase):

    def get_request(self, cookie=DEFAULT_COOKIE):
        # get a prepopulated test request
        request = Request.blank('http://localhost/test.html')
        if cookie is not None:
            request.headers['Cookie'] = cookie
        return request

    def test_plaintext_not_filtered(self):
        # plaintext docs are passed on
        app = ImageAdaptingMiddleware(wsgi_app_plaintext, {})
        request = self.get_request()
        response = request.get_response(app)
        self.assertEqual(
            response.headers['Content-Type'], 'text/plain; charset=UTF-8')
        self.assertEqual(
            response.body, 'Hi!')
        return

    def test_html_not_filtered(self):
        # html docs are passed on
        app = ImageAdaptingMiddleware(wsgi_app_html, {})
        request = self.get_request()
        response = request.get_response(app)
        self.assertEqual(
            response.headers['Content-Type'], 'text/html; charset=UTF-8')
        self.assertEqual(
            response.body, HTML)
        return

    def test_no_content_type(self):
        # requests without content type are ignored
        app = ImageAdaptingMiddleware(wsgi_app_no_content_type, {})
        request = self.get_request()
        response = request.get_response(app)
        self.assertEqual(
            response.headers['Content-Type'], '')
        self.assertEqual(
            response.body, b'')
        return

    def test_no_cookie(self):
        # ignore content without cookie
        app = ImageAdaptingMiddleware(wsgi_app_img_jpg, {})
        request = self.get_request(cookie=None)
        response = request.get_response(app)
        self.assertEqual(response.content_length, len(response.body))
        self.assertEqual(len(DATA_JPEG), len(response.body))
        return

    def test_invalid_cookie_key(self):
        # ignore if no 'resolution' cookie was sent
        app = ImageAdaptingMiddleware(wsgi_app_img_jpg, {})
        request = self.get_request(cookie='foo=bar; $Path=/')
        response = request.get_response(app)
        self.assertEqual(response.content_length, len(response.body))
        self.assertEqual(len(DATA_JPEG), len(response.body))
        return

    def test_invalid_cookie_value(self):
        # ignore non-integer cookie values
        app = ImageAdaptingMiddleware(wsgi_app_img_jpg, {})
        request = self.get_request(cookie='resolution=foo; $Path=/')
        response = request.get_response(app)
        self.assertEqual(response.content_length, len(response.body))
        self.assertEqual(len(DATA_JPEG), len(response.body))
        return

    def test_resolution_too_large(self):
        # do not change original if resolution > actual image size
        app = ImageAdaptingMiddleware(wsgi_app_img_jpg, {})
        request = self.get_request(cookie='resolution=512; $Path=/')
        response = request.get_response(app)
        self.assertEqual(response.content_length, len(response.body))
        self.assertEqual(len(DATA_JPEG), len(response.body))
        return

    def test_no_resolution_at_all(self):
        # do not change original if no resolution at all is given
        # or resolution < 1
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions='0')
        request = self.get_request(cookie='resolution=0; $Path=/')
        response = request.get_response(app)
        self.assertEqual(response.content_length, len(response.body))
        self.assertEqual(len(DATA_JPEG), len(response.body))
        return

    def test_shrink_jpeg_simple(self):
        # JPEGs are shrinked, if resolution is less than the original size
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions="128, 64, 32")
        request = self.get_request(cookie='resolution=62; $Path=/')
        response = request.get_response(app)
        img_data = response.body
        image = Image.open(StringIO(img_data))
        self.assertEqual(image.size, (64, 64))
        self.assertEqual(
            response.content_length, len(img_data))
        self.assertEqual(
            response.content_type, 'image/jpeg')
        return

    def test_shrink_jpeg_min(self):
        # JPEGs are shrinked to minimal size, if client width <= lowest resol.
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions="128, 64, 32")
        request = self.get_request(cookie='resolution=32; $Path=/')
        response = request.get_response(app)
        img_data = response.body
        image = Image.open(StringIO(img_data))
        self.assertEqual(image.size, (32, 32))
        return

    def test_shrink_jpeg_client_gt_image(self):
        # JPEGs are left untouched if client size > image size
        # This also applies, if the original size is not in set of resolutions
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions="64, 32")
        request = self.get_request(cookie='resolution=130; $Path=/')
        response = request.get_response(app)
        img_data = response.body
        image = Image.open(StringIO(img_data))
        self.assertEqual(image.size, (128, 128))
        return

    def test_non_mobile_client(self):
        # non mobile clients will get the largest resolution in
        # resolution list
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions="64, 32")
        request = self.get_request(cookie=None)
        request.headers['HTTP_USER_AGENT'] = 'amaya/9.51 libwww/5.4.0'
        response = request.get_response(app)
        img_data = response.body
        image = Image.open(StringIO(img_data))
        self.assertEqual(image.size, (64, 64))
        return

    def test_mobile_client(self):
        # mobile clients get the smallest resolution in resolution
        # list.
        app = ImageAdaptingMiddleware(
            wsgi_app_img_jpg, {}, resolutions="64, 32")
        request = self.get_request(cookie=None)
        request.headers[
            'HTTP_USER_AGENT'] = 'BlackBerry7730/3.7.1 UP.Link/5.1.2.5'
        response = request.get_response(app)
        img_data = response.body
        image = Image.open(StringIO(img_data))
        self.assertEqual(image.size, (32, 32))
        return


class GetClientResolutionTests(unittest.TestCase):
    # tests for ImageAdaptingMiddleware.get_client_resolution

    def setUp(self):
        self.request = Request.blank('http://localhost/test.html')
        self.middleware = ImageAdaptingMiddleware(None, {})
        return

    def test_no_cookie(self):
        # w/o cookie -> no resolution
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (None, 1))
        return

    def test_no_cookie_with_resolution_key(self):
        # we need a cookie named 'resolution' to get a resolution
        self.request.headers['Cookie'] = 'foo=bar; $Path=/'
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (None, 1))
        return

    def test_resolution_not_an_integer(self):
        # resolution cookie must contain some integer
        self.request.headers['Cookie'] = 'resolution=not-a-number; $Path=/'
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (None, 1))
        return

    def test_resolution(self):
        # integers can be extracted from the resolution cookie
        self.request.headers['Cookie'] = 'resolution=1024; $Path=/'
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (1024, 1))
        return

    def test_resolution_invalid_retina_value(self):
        # if retina value is broken, we get 1 as default
        self.request.headers['Cookie'] = 'resolution=123.trash; $Path=/'
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (123, 1))
        return

    def test_resolution_with_retina_value(self):
        # we can also get a retina value from the resolution cookie
        self.request.headers['Cookie'] = 'resolution=1024.2; $Path=/'
        self.assertEqual(
            self.middleware.get_client_resolution(self.request), (1024, 2))
        return


class ShouldIgnoreTests(unittest.TestCase):
    # tests for ImageAdaptingMiddleware.should_ignore

    def setUp(self):
        self.middleware = ImageAdaptingMiddleware(None, {})
        return

    def test_resolution_okay(self):
        # resolution values >= 1 are okay
        self.assertEqual(self.middleware.should_ignore(None, 123), False)
        return

    def test_resolution_too_small(self):
        # resolution values < 1 are not okay
        self.assertEqual(self.middleware.should_ignore(None, -12), True)
        return

    def test_resolution_none(self):
        # unset resolution values should be ignored
        self.assertEqual(self.middleware.should_ignore(None, None), True)
        return


class ShouldAdaptTests(unittest.TestCase):
    # tests for ImageAdaptingMiddleware.should_adapt

    def setUp(self):
        self.middleware = ImageAdaptingMiddleware(None, {})

    def test_no_content_type(self):
        # docs w/o content type should not be adapted
        request = Request.blank('http://localhost/test.html')
        response = request.get_response(wsgi_app_no_content_type)
        self.assertEqual(
            self.middleware.should_adapt(response), False)
        return

    def test_jpeg_content_type(self):
        # image/jpeg docs are adapted
        request = Request.blank('http://localhost/test.html')
        response = request.get_response(wsgi_app_img_jpg)
        self.assertEqual(
            self.middleware.should_adapt(response), True)
        return

    def test_gif_content_type(self):
        # image/gif docs are adapted
        request = Request.blank('http://localhost/test.html')
        response = request.get_response(wsgi_app_img_gif)
        self.assertEqual(
            self.middleware.should_adapt(response), True)
        return

    def test_png_content_type(self):
        # image/png docs are adapted
        request = Request.blank('http://localhost/test.html')
        response = request.get_response(wsgi_app_img_png)
        self.assertEqual(
            self.middleware.should_adapt(response), True)
        return


class IsMobileTests(unittest.TestCase):
    # tests for ImageAdaptingMiddleware.is_mobile()

    def setUp(self):
        self.middleware = ImageAdaptingMiddleware(None, {})
        self.request = Request.blank('http://localhost/test.html')

    def test_no_infos(self):
        # w/o any infos we assume a request comes from non-mobile
        self.assertEqual(
            False, self.middleware.is_mobile(self.request))
        return

    def test_non_mobile(self):
        # we can detect non-mobiles clients
        self.request.headers['HTTP_USER_AGENT'] = 'amaya/9.51 libwww/5.4.0'
        self.assertEqual(
            False, self.middleware.is_mobile(self.request))
        return

    def test_is_mobile(self):
        # we can detect mobile devices
        self.request.headers[
            'HTTP_USER_AGENT'] = 'BlackBerry7730/3.7.1 UP.Link/5.1.2.5'
        self.assertEqual(
            True, self.middleware.is_mobile(self.request))
        return


class InitParamsTests(unittest.TestCase):
    # Tests for initial values of ImageAdaptingMiddleware

    def test_resolutions_default(self):
        # we get a default for resolutions
        app = ImageAdaptingMiddleware(None, {})
        self.assertEqual(app.resolutions, [1382, 992, 768, 480])
        return

    def test_resolutions_liststring(self):
        # we can parse resolutions list from a string
        app = ImageAdaptingMiddleware(None, {}, resolutions='1, 2, 3')
        self.assertEqual(app.resolutions, [1, 2, 3])
        return

    def test_resolutions_single_value(self):
        # we can cope with a single value passed in
        app = ImageAdaptingMiddleware(None, {}, resolutions='666')
        self.assertEqual(app.resolutions, [666])
        return

    def test_resolutions_invalid(self):
        # invalid lists result in a value error
        self.assertRaises(
            ValueError,
            ImageAdaptingMiddleware,
            None, {}, resolutions='not-a-list-of-ints')
        return


class FactoryTests(unittest.TestCase):

    def test_filter_app(self):
        # we can get a factory for ImageAdaptingMiddleware
        app = filter_app('FakeApp', {'key': 'value'}, resolutions='128, 64')
        self.assertTrue(
            isinstance(app, ImageAdaptingMiddleware))
        return


pail/tests/test_helpers.py
import os
import shutil
import tempfile
import unittest
from PIL import Image
from pail.helpers import resize, get_file_length, to_int_list, get_resolution

sample_jpg = os.path.join(os.path.dirname(__file__), 'lena.jpg')
sample_png = os.path.join(os.path.dirname(__file__), 'lena.png')


class HelperTests(unittest.TestCase):

    def setUp(self):
        self.tempdir = tempfile.mkdtemp()

    def tearDown(self):
        shutil.rmtree(self.tempdir)

    def test_resize(self):
        # we can resize images
        im = Image.open(sample_jpg)
        self.assertEqual(im.size, (128, 128))
        im_type, result = resize(sample_jpg, 64)
        new_size = Image.open(result).size
        self.assertEqual(new_size, (64, 64))
        self.assertEqual(im_type, 'JPEG')
        return

    def test_resize_no_resolution(self):
        # w/o resolution, no resizing
        self.assertEqual(resize(sample_jpg, None), None)
        return

    def test_resize_too_small(self):
        # if resolution >= image width, we get no new image
        self.assertEqual(resize(sample_jpg, 128), None)
        self.assertEqual(resize(sample_jpg, 256), None)
        return

    def test_resize_invalid_img_path(self):
        # invalid paths result in `None
        self.assertEqual(
            resize('not-a-path', 64), None)
        return

    def test_resize_broken_file(self):
        # broken images are not resized
        broken = os.path.join(self.tempdir, 'broken.jpg')
        with open(broken, 'w') as fd:
            fd.write('blah blah')
        self.assertEqual(
            resize(broken, 64), None)
        return

    def test_get_file_length(self):
        # make sure get_file_length works as expected
        path = os.path.join(self.tempdir, 'testfile1')
        with open(path, 'w') as fd:
            fd.write('1234567890')
        self.assertEqual(
            get_file_length(open(path, 'rb')), 10)
        return

    def test_get_file_length_zero_length(self):
        # make sure get_file_length works as expected with zero length
        path = os.path.join(self.tempdir, 'testfile1')
        fd = open(path, 'w')
        fd.close()
        self.assertEqual(
            get_file_length(open(path, 'rb')), 0)
        return

    def test_to_int_list(self):
        # we can turn strings into lists of integers
        self.assertEqual(to_int_list('1, 2, 3'), [1, 2, 3])
        self.assertEqual(to_int_list('123'), [123])
        return

    def test_get_resolution_no_client_width(self):
        # mobiles get the smallest resolution, others the biggest
        self.assertEqual(
            get_resolution(None, 1, [480, 768, 922], True), 480)
        self.assertEqual(
            get_resolution(None, 1, [480, 768, 922], False), 922)
        return

    def test_get_resolution_pixel_density_eq_1(self):
        # get the next biggest value from resolution list
        self.assertEqual(
            get_resolution(480, 1, [480, 768, 922]), 480)
        self.assertEqual(
            get_resolution(512, 1, [480, 768, 922]), 768)
        self.assertEqual(
            get_resolution(800, 1, [480, 768, 922]), 922)
        self.assertEqual(
            get_resolution(1000, 1, [480, 768, 922]), 1000)
        return

    def test_get_resolution_pixel_density_ne_1(self):
        # with a pixel density > 1 we need larger images
        self.assertEqual(  # 2 x 480
            get_resolution(480, 2, [480, 768, 922]), 960)
        self.assertEqual(  # 2 x 768
            get_resolution(512, 2, [480, 768, 922]), 1536)
        self.assertEqual(  # 2 x 922
            get_resolution(800, 2, [480, 768, 922]), 1844)
        self.assertEqual(  # 2 x 922
            get_resolution(1000, 2, [480, 768, 922]), 2000)
        self.assertEqual(  # > 2 x 922
            get_resolution(2000, 2, [480, 768, 922]), 4000)
        return


pail/helpers.py
"""
Helper Functions
----------------

Minor helper functions.

"""
import tempfile
from PIL import Image


def resize(image, resolution=None, client_resolution=None, enlarge=False):
    """Resize `image` to have width `resolution`.

    `image` can be a path or some open file descriptor.

    `resolution` gives the desired maximum width in pixels.

    Returns ``None`` or a tuple

      ``(<FORMAT>, <RESULT_FILE_FD>)``

    with ``<FORMAT>`` being an image type as determined by PIL and
    ``<RESULT_FILE_FD>`` being an opened temporary file.

    If the image is already narrower than `resolution`, `None`
    is returned.

    The same applies if the image cannot be read or the resizing
    operation fails.
    """
    if resolution is None:
        return None
    try:
        im = Image.open(image)
    except IOError:
        return None
    size_x, size_y = im.size
    if size_x <= resolution:
        return None
    res_x = resolution
    res_y = resolution * size_y / size_x
    format = im.format

    #  create the resized version.
    new_im = im.resize((int(res_x), int(res_y)), Image.ANTIALIAS)

    #  save it to a temporary file.
    new_file = tempfile.TemporaryFile()
    new_im.save(new_file, format)
    new_file.seek(0)
    return format, new_file


def get_file_length(fd):
    """Get length of file denoted by a file descriptor.

    As we cannot get a stat of files of which we only have a file
    descriptor (and no path or similar), we have to do some tricks to
    get the file-length anyway.

    `fd` should be the descriptor of a file already open for reading.

    Returns the length of file as an integer.
    """
    fd.seek(0, 2)
    length = fd.tell()
    fd.seek(0)
    return length


def to_int_list(string):
    """Try to turn a string into a list of integers.

    Turns strings like `'1, 2, 3'` into regular lists like `[1, 2,
    3]`. The number have to be comma-separated. Also lists with only
    one entry are handled correctly.
    """
    if not ',' in string:
        return [int(string)]
    return [int(x) for x in string.split(',')]


def get_resolution(client_width, pixel_density, resolution_list,
                   is_mobile=True):
    """Get the desired resolution based on the input parameters.

    `client_width`
       The screen width of some client device.

    `pixel_density`
       By default 1. The amount of physical pixels representing a single
       CSS pixel. Normally 1 but on retina displays it can be more.

    `resolution_list`
       A list of supported screen widths (integers) from the web
       applications side.

    `is_mobile`
       A boolean. Indicates, whether the client device seems to be a
       mobile device. If so and no `client_width` is given, then the
       lowest possible resolution (based on `resolution_list`) is
       returned.
    """
    if client_width is None:
        return is_mobile and min(resolution_list) or max(resolution_list)
    possible_values = resolution_list
    possible_values.append(max( max(resolution_list), client_width))
    total_width = client_width * pixel_density
    result = [x * pixel_density for x in possible_values
              if total_width <= x * pixel_density]
    result = min(result + [max(possible_values) * pixel_density])
    return result


pail/wsgi.py
"""
WSGI Components
---------------

Middleware and other components that can be used with other WSGI
components.
"""
import re
import tempfile
from webob import Request
from webob.static import FileIter
from pail.helpers import resize, to_int_list, get_resolution

#: Default Resolution set used, when no other was given in config.
DEFAULT_RESOLUTIONS = "1382, 992, 768, 480"

#: Regular expressions from http://detectmobilebrowsers.com/ (public domain)
reg_b = re.compile(r"(android|bb\\d+|meego).+mobile|avantgo|bada\\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino", re.I|re.M)

reg_v = re.compile(r"1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\\-(n|u)|c55\\/|capi|ccwa|cdm\\-|cell|chtm|cldc|cmd\\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\\-s|devi|dica|dmob|do(c|p)o|ds(12|\\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\\-|_)|g1 u|g560|gene|gf\\-5|g\\-mo|go(\\.w|od)|gr(ad|un)|haie|hcit|hd\\-(m|p|t)|hei\\-|hi(pt|ta)|hp( i|ip)|hs\\-c|ht(c(\\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\\-(20|go|ma)|i230|iac( |\\-|\\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\\/)|klon|kpt |kwc\\-|kyo(c|k)|le(no|xi)|lg( g|\\/(k|l|u)|50|54|\\-[a-w])|libw|lynx|m1\\-w|m3ga|m50\\/|ma(te|ui|xo)|mc(01|21|ca)|m\\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\\-2|po(ck|rt|se)|prox|psio|pt\\-g|qa\\-a|qc(07|12|21|32|60|\\-[2-7]|i\\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\\-|oo|p\\-)|sdk\\/|se(c(\\-|0|1)|47|mc|nd|ri)|sgh\\-|shar|sie(\\-|m)|sk\\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\\-|v\\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\\-|tdg\\-|tel(i|m)|tim\\-|t\\-mo|to(pl|sh)|ts(70|m\\-|m3|m5)|tx\\-9|up(\\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\\-|your|zeto|zte\\-", re.I|re.M)


class ImageAdaptingMiddleware(object):
    """A WSGI middleware to shrink images on-the-fly.

    It works as a filter-wrapper that examines images delivered by
    other WSGI components and might reduce the image sizes on-the-fly.

    `resolutions`
       a string containing a comma separated list of integers. These
       number give the supported resolutions. The list does not have
       to be sorted.
    """

    #: The content types considered as handable. Only HTTP responses
    #: providing one of these types are handled by the middleware.
    acceptable_types = [
        'image/jpeg', 'image/png', 'image/gif']

    def __init__(self, app, global_conf, resolutions=DEFAULT_RESOLUTIONS):
        self.app = app
        self.global_conf = global_conf
        self.acceptable_types = [
            'image/jpeg', 'image/png', 'image/gif']
        try:
            self.resolutions = to_int_list(resolutions)
        except:
            raise ValueError('resolutions must be a comma-separated'
                             'list of integers')
        pass

    def __call__(self, environ, start_response):
        request = Request(environ)
        resolution = self.get_resolution(request)
        if self.should_ignore(request, resolution):
            return self.app(environ, start_response)
        response = request.get_response(self.app)

        if not self.should_adapt(response):
            return response(environ, start_response)

        new_img = self.create_resized_image(response, resolution)
        if new_img is None:
            return response(environ, start_response)

        im_type, new_img = new_img
        response.app_iter = FileIter(new_img)  # sets also content-length
        return response(environ, start_response)

    def create_resized_image(self, response, resolution):
        """Create a resized version of current content image.
        """
        orig_file = tempfile.TemporaryFile()
        for item in response.copy().app_iter:
            orig_file.write(item)
        orig_file.seek(0)
        return resize(orig_file, resolution)

    def get_client_resolution(self, request):
        """Get the client screen resolution from request.

        Returns a tuple ``(CLIENT_SCREEN_WIDTH, RETINA_VALUE)``.

        Here ``CLIENT_SCREEN_WITH`` gives the assumed screen width of the
        client device (an integer) or ``None``.

        ``RETINA_VALUE`` represents the factor of retina displays and
        is always an integer. If one CSS pixel stands for three
        physical pixels, this value would be ``3``. If no such value
        can be extracted from request, the default value ``1`` is
        returned.

        The method expects in `request` a cookie named `resolution`
        containing either a single integer or a float.

        If the value is an integer, we assume a retina value of
        ``1``. Otherwise we expect the client screen width before the
        dot and the retina factor after the dot.

        So a client providing a device width of 640 (CSS) pixels but
        with a retina display, that supports 3 physical pixels per CSS
        pixel, should pass in a cookie named `resolution` with value
        of ``640.3``.
        """
        cookies = request.cookies
        if not cookies:
            return (None, 1)
        cookie = cookies.get('resolution')
        if not cookie:
            return (None, 1)
        resolution, retina_value = cookie, '1'
        if '.' in cookie:
            resolution, retina_value = cookie.split('.', 1)
        try:
            resolution = int(resolution)
        except ValueError:
            resolution = None
        try:
            retina_value = int(retina_value)
        except ValueError:
            retina_value = 1
        return (resolution, retina_value)

    def get_resolution(self, request):
        """Determine a desired resolution from client screen
        resolution and available resolutions.
        """
        client_resolution, retina_value = self.get_client_resolution(request)
        is_mobile = self.is_mobile(request)
        return get_resolution(
            client_resolution, retina_value, self.resolutions, is_mobile)

    def should_ignore(self, request, resolution):
        """Should the given request be ignored?
        """
        if not resolution or resolution < 1:
            return True
        return False

    def should_adapt(self, response):
        """Should we adapt the response from the wrapped app?
        """
        if response.content_type not in self.acceptable_types:
            return False
        return True

    def is_mobile(self, request):
        """Is the device that sent `request` a mobile?
        """
        user_agent = request.headers.get('HTTP_USER_AGENT', None)
        if not user_agent:
            return False
        if reg_b.search(user_agent) or reg_v.search(user_agent[:4]):
            return True
        return False


def filter_app(app, global_conf, **kw):
    """A factory that returns `ImageAdaptingMiddleware` instances.
    """
    return ImageAdaptingMiddleware(app, global_conf, **kw)
